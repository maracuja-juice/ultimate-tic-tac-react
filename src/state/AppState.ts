import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
