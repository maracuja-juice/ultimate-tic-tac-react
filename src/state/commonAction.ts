export const RESTART_GAME = 'common/restart-game';

export const restartGame = () => ({
    type: RESTART_GAME
});