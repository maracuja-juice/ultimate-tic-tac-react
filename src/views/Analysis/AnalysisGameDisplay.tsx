import * as React from 'react';
import { BigBoard } from '../../components/Board/BigBoard/BigBoard';
import { AnalysisGame } from '../../state/AppState';
import './AnalysisGame.css';
import { HistoryButtons } from '../../components/Analysis/HistoryButtons/HistoryButtons';
import { MoveList } from '../../components/Analysis/MoveList/MoveList';
import { scroller } from 'react-scroll/modules';
import { moveScrollElementBaseName } from '../../components/Analysis/ScrollElementConstants';

interface AnalysisGameDisplayProps {
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
  analysisGame: AnalysisGame;
}

// TODO Hooks + Function Component?

export class AnalysisGameDisplay extends React.Component<
  AnalysisGameDisplayProps
> {
  scrollToElement = (moveNumberToScrollTo: number) => {
    scroller.scrollTo(moveScrollElementBaseName + moveNumberToScrollTo, {
      duration: 300,
      smooth: true,
      containerId: 'moveList',
      offset: -108,
    });
  };

  render() {
    const { analysisGame } = this.props;

    if (!analysisGame) {
      return <></>;
    }

    const reversedMoves = analysisGame.moves.slice().reverse();
    const currentlyAppliedMove =
      reversedMoves &&
      reversedMoves[
        reversedMoves.findIndex(
          (m) => m.moveNumber === analysisGame.currentMove,
        )!
      ];

    return (
      <div className="center">
        <div className="analysisLayout">
          <div id="moveList" className="moveList">
            <MoveList
              currentMove={analysisGame.currentMove}
              reversedMoves={reversedMoves}
              moveForwardInHistory={this.props.moveForwardInHistory}
              moveBackwardInHistory={this.props.moveBackwardInHistory}
            />
          </div>
          <div className="historyButtons">
            <HistoryButtons
              currentMove={analysisGame.currentMove}
              lastMove={reversedMoves[0] && reversedMoves[0].moveNumber}
              moveForwardInHistory={this.props.moveForwardInHistory}
              moveBackwardInHistory={this.props.moveBackwardInHistory}
              onInteraction={this.scrollToElement}
            />
          </div>
          <div className="analysisGame">
            <BigBoard
              // tslint:disable-next-line:no-empty
              onPlayerMoved={() => {}}
              board={analysisGame.board}
              activeBoards={analysisGame.activeBoards}
              currentPlayer={analysisGame.game.currentPlayer}
              markTileSpecially={{
                condition: currentlyAppliedMove !== undefined,
                position: currentlyAppliedMove
                  ? {
                      boardPosition: currentlyAppliedMove.boardPosition,
                      tilePosition: currentlyAppliedMove.tilePosition,
                    }
                  : undefined,
              }}
            />
          </div>
        </div>
      </div>
    );
  }
}
