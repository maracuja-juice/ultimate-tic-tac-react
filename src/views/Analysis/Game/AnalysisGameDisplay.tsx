import * as React from 'react';
import { BigBoard } from '../../../components/Board/BigBoard/BigBoard';
import { AnalysisGame } from '../../../state/AppState';
import { HistoryButtons } from '../../../components/Analysis/HistoryButtons/HistoryButtons';
import { MoveList } from '../../../components/Analysis/MoveList/MoveList';
import { scroller } from 'react-scroll/modules';
import { moveScrollElementBaseName } from '../../../components/Analysis/ScrollElementConstants';
import styles from './AnalysisGame.module.css';

interface AnalysisGameDisplayProps {
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
  analysisGame: AnalysisGame;
}

function scrollToElement(moveNumberToScrollTo: number) {
  scroller.scrollTo(moveScrollElementBaseName + moveNumberToScrollTo, {
    duration: 300,
    smooth: true,
    containerId: 'moveList',
    // Thanks to this, when scrolling to an element, some elements above it will also be displayed
    offset: -125,
  });
}

export function AnalysisGameDisplay(props: AnalysisGameDisplayProps) {
  const analysisGame = props.analysisGame;
  const reversedMoves = analysisGame.moves.slice().reverse();
  const currentlyAppliedMove =
    reversedMoves[
      reversedMoves.findIndex((m) => m.moveNumber === analysisGame.currentMove)!
    ];

  return (
    <div className={styles.analysisLayout}>
      <div id="moveList" className={styles.moveList}>
        <MoveList
          currentMove={analysisGame.currentMove}
          reversedMoves={reversedMoves}
          moveForwardInHistory={props.moveForwardInHistory}
          moveBackwardInHistory={props.moveBackwardInHistory}
        />
      </div>
      <div className={styles.historyButtons}>
        <HistoryButtons
          currentMove={analysisGame.currentMove}
          lastMove={reversedMoves[0] && reversedMoves[0].moveNumber}
          moveForwardInHistory={props.moveForwardInHistory}
          moveBackwardInHistory={props.moveBackwardInHistory}
          onInteraction={scrollToElement}
        />
      </div>
      <div className={styles.analysisGame}>
        <BigBoard
          board={analysisGame.board}
          activeBoards={analysisGame.activeBoards}
          currentPlayer={analysisGame.game.currentPlayer}
          markTileSpecially={{
            condition: currentlyAppliedMove !== undefined,
            position: currentlyAppliedMove
              ? {
                  boardPosition: currentlyAppliedMove.boardPosition,
                  tilePosition: currentlyAppliedMove.tilePosition,
                }
              : undefined,
          }}
        />
      </div>
    </div>
  );
}
