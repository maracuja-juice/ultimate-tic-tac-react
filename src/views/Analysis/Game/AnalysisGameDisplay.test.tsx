import * as React from 'react';
import { shallow } from 'enzyme';
import {
  circleFinishedBoardMock,
  movesForCircleFinishedBoardMock,
} from '../../../__mocks__';
import { Player, Winner } from '../../../state/AppState';
import { AnalysisGameDisplay } from './AnalysisGameDisplay';

// tslint:disable:no-empty
describe('AnalysisGame', () => {
  it('should match snapshot', () => {
    const moveForwardInHistory = jest.fn((numberOfMoves) => {});
    const moveBackwardInHistory = jest.fn((numberOfMoves) => {});

    const analysis = shallow(
      <AnalysisGameDisplay
        moveForwardInHistory={moveForwardInHistory}
        moveBackwardInHistory={moveBackwardInHistory}
        analysisGame={{
          board: circleFinishedBoardMock,
          activeBoards: [],
          currentMove:
            movesForCircleFinishedBoardMock[
              movesForCircleFinishedBoardMock.length - 1
            ].moveNumber,
          game: {
            currentPlayer: Player.Cross,
            isFinished: true,
            winningPlayer: Winner.Circle,
          },
          moves: movesForCircleFinishedBoardMock,
          id: '1',
        }}
      />,
    );

    expect(analysis).toMatchSnapshot();
  });
});
