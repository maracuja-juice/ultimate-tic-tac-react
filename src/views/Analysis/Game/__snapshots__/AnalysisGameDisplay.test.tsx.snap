// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalysisGame should match snapshot 1`] = `
<div
  className="centerHorizontal applyHeight"
>
  <div
    className="analysisLayout"
  >
    <div
      className="moveList"
      id="moveList"
    >
      <MoveList
        currentMove={30}
        moveBackwardInHistory={[MockFunction]}
        moveForwardInHistory={[MockFunction]}
        reversedMoves={
          Array [
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 30,
              "player": 1,
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 0,
              },
              "moveNumber": 29,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 1,
              },
              "moveNumber": 28,
              "player": 1,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 2,
              },
              "moveNumber": 27,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 1,
              },
              "moveNumber": 26,
              "player": 1,
              "tilePosition": Object {
                "x": 2,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 2,
              },
              "moveNumber": 25,
              "player": 0,
              "tilePosition": Object {
                "x": 1,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 1,
              },
              "moveNumber": 24,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 2,
              },
              "moveNumber": 23,
              "player": 0,
              "tilePosition": Object {
                "x": 1,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 22,
              "player": 1,
              "tilePosition": Object {
                "x": 2,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 1,
              },
              "moveNumber": 21,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 0,
              },
              "moveNumber": 20,
              "player": 1,
              "tilePosition": Object {
                "x": 2,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 1,
              },
              "moveNumber": 19,
              "player": 0,
              "tilePosition": Object {
                "x": 2,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 18,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 1,
              },
              "moveNumber": 17,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 16,
              "player": 1,
              "tilePosition": Object {
                "x": 2,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 2,
              },
              "moveNumber": 15,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 1,
              },
              "moveNumber": 14,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 1,
              },
              "moveNumber": 13,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 1,
              },
              "moveNumber": 12,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 1,
              },
              "moveNumber": 11,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 2,
              },
              "moveNumber": 10,
              "player": 1,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 2,
              },
              "moveNumber": 9,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 2,
              },
              "moveNumber": 8,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 2,
                "y": 2,
              },
              "moveNumber": 7,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 2,
              },
              "moveNumber": 6,
              "player": 1,
              "tilePosition": Object {
                "x": 2,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 0,
              },
              "moveNumber": 5,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 1,
              },
              "moveNumber": 4,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 0,
              },
              "moveNumber": 3,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 2,
              "player": 1,
              "tilePosition": Object {
                "x": 1,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 1,
                "y": 0,
              },
              "moveNumber": 1,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 0,
              },
            },
          ]
        }
      />
    </div>
    <div
      className="historyButtons"
    >
      <HistoryButtons
        currentMove={30}
        lastMove={30}
        moveBackwardInHistory={[MockFunction]}
        moveForwardInHistory={[MockFunction]}
        onInteraction={[Function]}
      />
    </div>
    <div
      className="analysisGame"
    >
      <BigBoard
        activeBoards={Array []}
        board={
          Array [
            Object {
              "position": Object {
                "x": 0,
                "y": 0,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 1,
                },
              ],
              "value": 1,
            },
            Object {
              "position": Object {
                "x": 0,
                "y": 1,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 2,
                },
              ],
              "value": 1,
            },
            Object {
              "position": Object {
                "x": 0,
                "y": 2,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 1,
                },
              ],
              "value": 1,
            },
            Object {
              "position": Object {
                "x": 1,
                "y": 0,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 2,
                },
              ],
              "value": 0,
            },
            Object {
              "position": Object {
                "x": 1,
                "y": 1,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 1,
                },
              ],
              "value": 1,
            },
            Object {
              "position": Object {
                "x": 1,
                "y": 2,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 2,
                },
              ],
              "value": 2,
            },
            Object {
              "position": Object {
                "x": 2,
                "y": 0,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 1,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 2,
                },
              ],
              "value": 2,
            },
            Object {
              "position": Object {
                "x": 2,
                "y": 1,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 2,
                },
              ],
              "value": 0,
            },
            Object {
              "position": Object {
                "x": 2,
                "y": 2,
              },
              "tiles": Array [
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 1,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 0,
                    "y": 2,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 1,
                  },
                  "value": 0,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 1,
                    "y": 2,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 0,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 1,
                  },
                  "value": 2,
                },
                Object {
                  "boardPosition": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "position": Object {
                    "x": 2,
                    "y": 2,
                  },
                  "value": 2,
                },
              ],
              "value": 2,
            },
          ]
        }
        currentPlayer={0}
        markTileSpecially={
          Object {
            "condition": true,
            "position": Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
          }
        }
        onPlayerMoved={[Function]}
      />
    </div>
  </div>
</div>
`;
